File: app/src/main/java/com/example/smarthome/ligths/ui/AddLightFragment.java
package com.example.smarthome.ligths.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.example.smarthome.R;
import com.example.smarthome.ligths.viewModels.LightViewModel;
import com.google.android.material.switchmaterial.SwitchMaterial;

public class AddLightFragment extends Fragment {

    private LightViewModel model;

    public AddLightFragment(){}

    public static AddLightFragment newInstance() {

        Bundle args = new Bundle();

        AddLightFragment fragment = new AddLightFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        model = new ViewModelProvider(requireActivity()).get(LightViewModel.class);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.add_light_fragment, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        final EditText etName = view.findViewById(R.id.light_add_et_name);
        final EditText etIp = view.findViewById(R.id.light_add_et_ip);
        final SwitchMaterial switchOnOf = view.findViewById(R.id.light_add_switch_onof);
        final Button bAdd = view.findViewById(R.id.light_add_button_Add);

        bAdd.setOnClickListener(v -> {
            String name = etName.getText().toString();
            String ip = etIp.getText().toString();
            boolean on = switchOnOf.isChecked();

            //todo send light to server
        });
    }
}

File: app/src/main/java/com/example/smarthome/ui/light/AddLightFragment.kt
package com.example.smarthome.ui.light

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import com.example.smarthome.R
import com.example.smarthome.data.model.Light
import com.example.smarthome.databinding.AddLightFragmentBinding
import com.example.smarthome.utilities.LiveDataObservers
import com.example.smarthome.utilities.OnClickListeners
import com.example.smarthome.utilities.Resource
import com.example.smarthome.utilities.ViewHelper
import com.example.smarthome.viewmodel.LightViewModel
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class AddLightFragment :
    Fragment(R.layout.add_light_fragment),
    LiveDataObservers,
    OnClickListeners,
    ViewHelper
{
    private var _binding: AddLightFragmentBinding? = null
    private val binding get() = _binding!!
    private val lightViewModel: LightViewModel by viewModels()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = AddLightFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initLiveDataObservers()
        initOnClickListeners()
    }

    override fun initLiveDataObservers() {
        lightViewModel.status.observe(viewLifecycleOwner) {
            when (it.status) {
                Resource.Status.LOADING -> setLoadingLayout()
                Resource.Status.ADDED -> requireActivity().finish()
                else -> setNormalLayout()
            }

            toast(requireActivity(), it.message)
        }
    }

    override fun initOnClickListeners() {
        binding.addButton.setOnClickListener {
            val name: String = binding.name.text.toString()
            val mac: String = binding.mac.text.toString()
            val isOn: Boolean = binding.switchButton.isChecked

            lightViewModel.add(Light(name, mac, isOn))
        }
    }

    private fun setLoadingLayout() {
        with (binding) {
            showViews(progressBar)
            hideViews(addButton)
        }
    }

    private fun setNormalLayout() {
        with (binding) {
            hideViews(progressBar)
            showViews(addButton)
        }
    }
}